lib/streamlit/runtime/partials.py

```diff
@@ -0,0 +1,196 @@
+# Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022)
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+import hashlib
+import pickle
+from functools import wraps
+from typing import (
+    Any,
+    Callable,
+    Iterator,
+    MutableMapping,
+    Optional,
+    TypeVar,
+    Union,
+    overload,
+)
+
+from streamlit.proto.ForwardMsg_pb2 import ForwardMsg
+from streamlit.runtime.scriptrunner import get_script_run_ctx
+from streamlit.runtime.state.session_state_proxy import get_session_state
+
+F = TypeVar("F", bound=Callable[..., Any])
+WrappedPartial = Callable[[], Any]
+
+
+@overload
+def partial(
+    func: F,
+    *,
+    run_every: Optional[float] = None,
+) -> F:
+    ...
+
+
+@overload
+def partial(
+    func: None = None,
+    *,
+    run_every: Optional[float] = None,
+) -> Callable[[F], F]:
+    ...
+
+
+def partial(
+    func: Optional[F] = None,
+    *,
+    run_every: Optional[float] = None,
+) -> Union[Callable[[F], F], F]:
+    if func is None:
+        # Support passing the params via function decorator
+        def wrapper(f: F) -> F:
+            return partial(
+                func=f,
+                run_every=run_every,
+            )
+
+        return wrapper
+    else:
+        # To make mypy type narrow Optional[F] -> F
+        non_optional_func = func
+
+    print(non_optional_func.__qualname__)
+
+    @wraps(non_optional_func)
+    def wrap(*args, **kwargs):
+        import streamlit as st
+
+        ctx = get_script_run_ctx()
+        if ctx is None:
+            return
+
+        # HACK: FIXME FIXME FIXME FIXME FIXME FIXME FIXME FIXME FIXME FIXME FIXME FIXME
+        # The serialization/deserialization of the dg_stack here is essentially just
+        # being used to reset the stack before each partial rerun. This is necessary
+        # because dg_stack is mutated as the script runs, but we need to save its state
+        # at this point so that we can rewrite a specific part of the app in a partial
+        # rerun. We'll eventually want to make changes to the DeltaGenerator class
+        # itself to support this in a less hacky way.
+        if not len(ctx.dg_stack):
+            with st.container():
+                dg_stack = pickle.dumps(ctx.dg_stack)
+                active_dg = ctx.dg_stack[-1]
+        else:
+            dg_stack = pickle.dumps(ctx.dg_stack)
+            active_dg = ctx.dg_stack[-1]
```

Comment:
Per the discussion today, I think we want it to never do this - right? We always want it to be wrapped with its own container?